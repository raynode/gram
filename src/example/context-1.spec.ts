import {
  graphql,
  GraphQLBoolean,
  GraphQLInt,
  GraphQLString,
  introspectionQuery,
  printSchema,
} from 'graphql'
import { createSchemaBuilder, NodeType, Service } from '..'

interface User extends NodeType {
  email: string
}

interface UserAction extends NodeType {
  userId: string
}

type SchemaTypes = 'admin' | 'user'

describe('testing the example 1', () => {
  it('should build the example code', async () => {
    interface GQLContext {
      authToken?: string
      authId?: string
    }

    const builder = createSchemaBuilder<SchemaTypes, GQLContext>()
    const user = builder.model('User')
    user.attr('email', GraphQLString)

    builder.addQuery(({ buildMode }) => ({
      name: 'buildMode',
      type: GraphQLString,
      resolver: () => buildMode,
    }))

    builder.addQuery(({ buildMode }) => ({
      name: 'me',
      type: user,
      resolver: (root, args, queryContext) => {
        if (!queryContext) throw new Error('Need an authToken-context')
        if (buildMode === 'user' && !queryContext.authToken)
          throw new Error('Need an user:authToken')
        if (!queryContext.authId) throw new Error('Need an authID')
        return {
          id: queryContext.authId,
          email: buildMode === 'admin' ? 'admin' : queryContext.authToken,
        }
      },
    }))

    const adminSchema = builder.build('admin')
    const userSchema = builder.build('user')

    const source = `{ buildMode me { email }}`

    const adminResult = await graphql({
      schema: adminSchema,
      source,
      contextValue: {
        authId: 'AdminID',
      },
    })
    const userResult = await graphql({
      schema: userSchema,
      source,
      contextValue: {
        authId: 'AuthenticationID',
        authToken: 'My@Token.com',
      },
    })
    expect(printSchema(adminSchema)).toMatchSnapshot()
    expect(printSchema(userSchema)).toMatchSnapshot()
    expect(adminResult).not.toHaveProperty('errors')
    expect(userResult).not.toHaveProperty('errors')
    expect(userResult.data).toEqual({
      buildMode: 'user',
      me: { email: 'My@Token.com' },
    })
    expect(adminResult.data).toEqual({
      buildMode: 'admin',
      me: { email: 'admin' },
    })
  })
})
