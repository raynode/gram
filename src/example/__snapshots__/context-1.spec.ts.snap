// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the example 1 should build the example code 1`] = `
"input CreateUserData {
  email: String
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

scalar DateTime

interface List {
  page: Page
  nodes: [Node!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  buildMode: String
  me: User
}

input UpdateUserData {
  email: String
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type User implements Node {
  email: String
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input UserFilter {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  email: String
  email_not: String
  email_in: [String!]
  email_not_in: [String!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input UserPage {
  limit: Int
  offset: Int
}

type Users implements List {
  page: Page
  nodes: [User!]!
}

enum UserSortOrder {
  email_ASC
  email_DESC
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
}

input UserWhere {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  email: String
  email_not: String
  email_in: [String!]
  email_not_in: [String!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}
"
`;

exports[`testing the example 1 should build the example code 2`] = `
"input CreateUserData {
  email: String
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

scalar DateTime

interface List {
  page: Page
  nodes: [Node!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  buildMode: String
  me: User
}

input UpdateUserData {
  email: String
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type User implements Node {
  email: String
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input UserFilter {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  email: String
  email_not: String
  email_in: [String!]
  email_not_in: [String!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input UserPage {
  limit: Int
  offset: Int
}

type Users implements List {
  page: Page
  nodes: [User!]!
}

enum UserSortOrder {
  email_ASC
  email_DESC
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
}

input UserWhere {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  email: String
  email_not: String
  email_in: [String!]
  email_not_in: [String!]
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}
"
`;
